### PR TITLE
Add streaming calculation of ed25519

### PR DESCRIPTION
**Pull request by popzxc**
This PR allows to calculate `ed25519` via `State` object.

### CODE REVIEW

**Comment by popzxc** [comment]
Hi! Could you please clarify if there is something wrong with this PR? Or when will it be merged in other case?

**Comment by Dylan-DPC** [comment]
@popzxc Hi. Sorry about the delays. We will review this as soon as possible.

**Comment by popzxc** [comment]
@Dylan-DPC hello! Could you please clarify which tests should I add to get this PR merged?

**Comment by Dylan-DPC** [comment]
@popzxc you need to add test vectors from libsodium that are related to your content.

**Comment by kpp** [comment]
Hi. Need help?

**Comment by popzxc** [comment]
@kpp Hello. That would be really appreciated, since I'm busy on my main work project and don't know when I'll be free enough to complete this PR.

**Comment by kpp** [comment]
@Dylan-DPC I can't find any test in libsodium for `crypto_sign_ed25519ph_*` functions. Which tests do you want to be included?

**Comment by kpp** [comment]
The coverage for sign/ed25519.rs is 96.9%. Missing coverage:

```
  152  #[derive(Copy, Clone)] 
  153  pub struct State(ffi::crypto_sign_ed25519ph_state); 

  198  impl fmt::Debug for State { 
  199      fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { 
  200          write!(f, "ed25519 state") 
  201      } 
  202  } 
  203   
  204  // Impl Default becuase `State` does have a sensible default: State::init() 
  205  impl Default for State { 
  206      fn default() -> State { 
  207          State::init() 
  208      } 
  209  } 
  210   
```

**Comment by kpp** [comment]
@popzxc see coverage report. Will you fix missing coverage?

**Comment by kpp** [comment]
> I've added more tests: for missing coverage, as well as test vectors.

Give me some time, because there is no coverage report (I don't know why). So I will check it manually

**Comment by kpp** [comment]
Well, it is a 100% coverage.

@Dylan-DPC Any other notes?

**Comment by Dylan-DPC** [comment]
Would like a review from @dnaq

**Comment by dnaq** [comment]
This weekend at earliest

**Comment by kpp** [comment]
@dnaq ping

**Comment by dnaq** [comment]
Hopefully this weekend

**Comment by popzxc** [comment]
Hello! Is there something going on with this PR? :)

**Comment by kpp** [comment]
Woah. Sorry, I forgot about it

File: src/crypto/sign/ed25519.rs [hunk 1]
```diff
@@ -56,99 +58,167 @@ new_type! {
...
+    pub fn update(&mut self, m: &[u8]) {
+        unsafe {
+            ffi::crypto_sign_ed25519ph_update(&mut self.0, m.as_ptr(), m.len() as c_ulonglong);
+        }
+    }
+
+    /// `finalize()` finalizes the hashing computation and returns a `Signature`.
+    // Moves self becuase libsodium says the state should not be used
+    // anymore after final().
+    pub fn finalize(&mut self, &SecretKey(ref sk): &SecretKey) -> Signature {
```
Comment (kpp) [inline src/crypto/sign/ed25519.rs]: >     // Moves self becuase libsodium says the state should not be used anymore after final().

Then you should use self by value, not by `&mut self`

**Comment by kpp** [comment]
\+ add test with `finalize` without `update` right after `init`

**Comment by popzxc** [comment]
Done

**Comment by Dylan-DPC** [comment]
bors: r+

**Pull request merged**